wire-runtime/src/commonMain/kotlin/com/squareup/wire/Message.kt

```diff
@@ -30,7 +30,7 @@ expect abstract class Message<M : Message<M, B>, B : Message.Builder<M, B>> prot
    * Returns a byte string containing the proto encoding of this message's unknown fields. Returns
    * an empty byte string if this message has no unknown fields.
    */
-  fun unknownFields(): ByteString
+  open val unknownFields: ByteString
```

Comment:
Combine with ctor param?

Reply:
Why open? Can't subtypes just pass their value to this supertype?

Reply:
1. `Expected class constructor can't have a property parameter`.
2. `data` classes can't have non-property parameters in the primary constructor, so they have to declare a `val unknownFields`. The only way to avoid it clashing with the supertype's `val` is to `override`, which requires `open` or `abstract`. Kotlin doesn't allow declaring `abstract` primary constructor properties, so `open` is the only thing that worked for me. Agree that it sucks, is there another way to satisfy these constraints that I'm missing?

Reply:
Makes sense. Seems like we should nuke the base class constructor param then and just declare it abstract.

Reply:
That could work, but we'll need to add a new constructor to preserve binary compat. I'll give it a try.

Reply:
Nah, making the property abstract becomes a PITA for Java backwards compat. Happy to either merge as is, or punt until post-3.0.

Reply:
The problem is that right now you store the value in two fields: one in the supertype and one in the subtype.

Reply:
Yeah, good point. I can't think of different solution that would not require regenerating Java code.

Reply:
Me neither.

You could remove the `data` modifier and generate your own `equals`/`hashCode`/`copy`. You'd wind up making the classes smaller since the `componentN` functions probably aren't super useful since proto allows field re-ordering (provided the tag stays the same).